src/main/java/org/apache/commons/codec/binary/BaseNCodecInputStream.java

```diff
@@ -171,7 +174,7 @@ public int read(final byte array[], final int offset, final int len) throws IOEx
         */
         while (readLen == 0) {
```

Comment:
You commented that the read only attempts to read at most the size of the input buffer (`buf`), effectively ignoring the user requested length to read (`len`). This prevents a user reading with a very large `array` in a single method call.

How about this:
```java
// Attempt to read the request length
while (readLen < len) {
    if (!baseNCodec.hasData(context)) {
        // Obtain more data.
        // *** TBC: remove buffer allocation here ***
        final byte[] buf = new byte[doEncode ? 4096 : 8192];
        final int c = in.read(buf);
        if (doEncode) {
            baseNCodec.encode(buf, 0, c, context);
        } else {
            baseNCodec.decode(buf, 0, c, context);
        }
    }
    final int read = baseNCodec.readResults(array, offset + readLen, len - readLen, context);
    if (read < 0) {
        // Return the amount read or EOF
        return readLen != 0 ? readLen : -1;
    }
    readLen += read;
}
```

The above change passes the test suite using the old code. Can you try it with your changes to recycle the buffer.

IIUC recycling the buffer will not cause memory expansion here. With the new changes the call to `readResults` should either: read all the data if the user has provided a very large destination `array` and the Context is reset to zero; or read some of the data up to the `len` requested by the user. The Context will then return true for the next call into this method in the `hasData(context)` statement and the Context will be emptied next time.

Reply:
Looks good. Tests pass. I adjusted the new Base64-Test to provided bigger buffers.